File: java/TimeZones.java
package org.libphonenumber.ae;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Timezone lookup — the idiomatic Java surface over the shared Aether engine's
 * {@code PhoneNumberToTimeZonesMapper} side-library (ABI v5).
 *
 * <p>A longest-prefix match over the number's E.164 digits. Every call takes a
 * raw {@code (region, input)} like the rest of the binding; the engine parses to
 * E.164 internally. The unknown-zone sentinel is {@code "Etc/Unknown"}.
 *
 * <pre>{@code
 * TimeZones.timeZonesForNumber("US", "2015550123");  // ["America/New_York"]
 * TimeZones.timeZonesForNumber("GB", "2070313000");  // ["Europe/London"]
 * TimeZones.unknownTimeZone();                       // "Etc/Unknown"
 * }</pre>
 *
 * <p>Carries <b>no phone-number logic</b>: every method marshals to an
 * {@code aether_pn_embed_tz_*} call in {@link Native}. All methods are static
 * and stateless; the engine is loaded lazily and cached via {@link Native#load}.
 *
 * <p>Requires {@code --enable-native-access=ALL-UNNAMED} on the command line.
 */
public final class TimeZones {

    private TimeZones() {
    }

    private static Native api() {
        return Native.load(null);
    }

    /**
     * The IANA timezone ids for a number, as a list. A number with no known
     * zone maps to a single-element list of {@link #unknownTimeZone()}.
     */
    public static List<String> timeZonesForNumber(String region, String input) {
        Native a = api();
        int n = a.calli2s(a.tzCount, region, input);
        if (n == 0) {
            return Collections.singletonList(unknownTimeZone());
        }
        List<String> out = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            out.add(a.call2s1i(a.tzAt, region, input, i));
        }
        return Collections.unmodifiableList(out);
    }

    /** The number of zones for the number ({@code 0} = only the unknown zone). */
    public static int timeZoneCount(String region, String input) {
        Native a = api();
        return a.calli2s(a.tzCount, region, input);
    }

    /** The unknown-zone sentinel, {@code "Etc/Unknown"}. */
    public static String unknownTimeZone() {
        Native a = api();
        try {
            return a.takeString(
                    (java.lang.foreign.MemorySegment) a.tzUnknown.invokeExact());
        } catch (Throwable t) {
            throw Native.wrap(t);
        }
    }
}

File: php/src/TimeZones.php
<?php

/**
 * IANA time-zone lookup over the phonenumber engine (ABI v5).
 *
 * Mirrors libphonenumber's `PhoneNumberToTimeZonesMapper`. Carries no
 * phone-number logic — every method marshals to an `aether_pn_embed_tz_*` call
 * in {@see Native}. The unknown-zone sentinel is `"Etc/Unknown"`.
 *
 *     use PhoneNumberAe\TimeZones;
 *
 *     TimeZones::timeZonesForNumber('US', '2015550123'); // ['America/New_York']
 *     TimeZones::timeZonesForNumber('GB', '2070313000'); // ['Europe/London']
 *     TimeZones::unknownTimeZone();                      // 'Etc/Unknown'
 *
 * @package PhoneNumberAe
 */

declare(strict_types=1);

namespace PhoneNumberAe;

use FFI;

final class TimeZones
{
    /** This class is never instantiated — the stateless ABI has no state to hold. */
    private function __construct()
    {
    }

    private static function ffi(): FFI
    {
        return Native::load();
    }

    /**
     * The IANA time-zone ids for $input in $region, in order.
     *
     * A number with no known zone maps to a single-element list of the unknown
     * zone (`['Etc/Unknown']`).
     *
     * @return list<string>
     */
    public static function timeZonesForNumber(string $region, string $input): array
    {
        $ffi = self::ffi();
        $count = $ffi->aether_pn_embed_tz_count($region, $input);
        if ($count === 0) {
            return [self::unknownTimeZone()];
        }
        $out = [];
        for ($i = 0; $i < $count; $i++) {
            $out[] = Native::takeString($ffi, $ffi->aether_pn_embed_tz_at($region, $input, $i));
        }
        return $out;
    }

    /** How many time zones $input maps to in $region (0 = only the unknown zone). */
    public static function timeZoneCount(string $region, string $input): int
    {
        return self::ffi()->aether_pn_embed_tz_count($region, $input);
    }

    /** The unknown-zone sentinel, "Etc/Unknown". */
    public static function unknownTimeZone(): string
    {
        $ffi = self::ffi();
        return Native::takeString($ffi, $ffi->aether_pn_embed_tz_unknown());
    }
}
